File: Parallel-Csharp/src/Parallel/MyRegex/MyRegex.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Parallel_Csharp.Parallel.MyRegex;

public class MyRegex
{
    private int files = 0, ok = 0, warns = 0, css = 0, errs = 0, overflow = 0;
    public void Scan(List<string> lines, Regex[] patterns, SortedDictionary<Regex, int> counts)
    {
        Regex okPattern = 
            new Regex("No warnings or errors were found.", 
                RegexOptions.Compiled);
        Regex countPattern =
            new Regex("([0-9]+) warnings, ([0-9]+) errors were found!.*?(Not all warnings/errors were shown.)?",
                RegexOptions.Compiled);
        Regex countPattern2 =
            new Regex(
                "Tidy found ([0-9]+) warning[s]? and ([0-9]+) error[s]?!.*?(Not all warnings/errors were shown.)?",
                RegexOptions.Compiled);
        Regex cssPattern = new Regex("You are recommended to use CSS.*", RegexOptions.Compiled);
        Regex guardPattern = new Regex("line [0-9]+ column [0-9]+ - (Error|Warning):.*", RegexOptions.Compiled);
        Match m;
        files++;
        foreach (string line in lines)
        {
            if (okPattern.IsMatch(line))
            {
                ok++;
            }
            else if ((m = countPattern.Match(line)).Success)
            {
                warns += Convert.ToInt32(m.Groups[1].Value);
                errs += Convert.ToInt32(m.Groups[2].Value);
                if (m.Groups[3].Success)
                    overflow++;
            }
            else if ((m = countPattern2.Match(line)).Success)
            {
                warns += Convert.ToInt32(m.Groups[1].Value);
                errs += Convert.ToInt32(m.Groups[2].Value);
                if (m.Groups[3].Success)
                    overflow++;
            }
            else if ((m = guardPattern.Match(line)).Success)
            {
                bool found = false;
                foreach (Regex p in patterns)
                {
                    if ((m = p.Match(line)).Success)
                    {
                        found = true;
                        int i = counts.GetValueOrDefault(p, 0);
                        counts[p] = i + 1;
                        break;
                    }
                }

                if (!found)
                    Console.Error.WriteLine("Unrecognized line: " + line);
            }
            else if (cssPattern.Match(line).Success)
            {
                css++;
            }
        }
    }
}

File: Parallel-Java/src/parallel/myregex/MyRegex.java
package parallel.myregex;

import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MyRegex {
    void scan(List<String> lines, Pattern[] patterns, Map<Pattern, Integer> counts) {
        int files = 0, ok = 0, warns = 0, css = 0, errs = 0, overflow=0;

        Pattern okPattern = Pattern.compile("No warnings or errors were found.");
        Pattern countPattern = Pattern.compile("([0-9]+) warnings, ([0-9]+) errors were found!.*?(Not all warnings/errors were shown.)?");
        Pattern countPattern2 = Pattern.compile("Tidy found ([0-9]+) warning[s]? and ([0-9]+) error[s]?!.*?(Not all warnings/errors were shown.)?");
        Pattern cssPattern = Pattern.compile("You are recommended to use CSS.*");
        Pattern guardPattern = Pattern.compile("line [0-9]+ column [0-9]+ - (Error|Warning):.*");
        Matcher m;
        files++;
        for (String line: lines) {
            if (okPattern.matcher(line).matches()) {
                ok++;
            } else if ((m = countPattern.matcher(line)).matches()) {
                warns += Integer.parseInt(m.group(1));
                errs += Integer.parseInt(m.group(2));
                if (m.group(3) != null)
                    overflow++;
            } else if ((m = countPattern2.matcher(line)).matches()) {
                warns += Integer.parseInt(m.group(1));
                errs += Integer.parseInt(m.group(2));
                if (m.group(3) != null)
                    overflow++;
            } else if ((m = guardPattern.matcher(line)).matches()) {
                boolean found = false;
                for (Pattern p: patterns) {
                    if ((m = p.matcher(line)).matches()) {
                        found = true;
                        Integer i = counts.get(p);
                        counts.put(p, (i == null) ? 1 : i + 1);
                        break;
                    }
                }
                if (!found)
                    System.err.println("Unrecognized line: " + line);
            } else if (cssPattern.matcher(line).matches()) {
                css++;
            }
        }
    }
}
